sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala

```diff
@@ -456,11 +456,25 @@ final class DataFrameNaFunctions private[sql](df: DataFrame) {
     val keyExpr = df.col(col.name).expr
     def buildExpr(v: Any) = Cast(Literal(v), keyExpr.dataType)
     val branches = replacementMap.flatMap { case (source, target) =>
-      Seq(buildExpr(source), buildExpr(target))
+        if (!(checkIfNonFractionalNumeric(col.dataType) &&
```

Comment:
You could do `col.dataType match` instead of the separate function with the long boolean expression there. I think that's more idiomatic to Scala and in the context of this part of the code.